plugins/sla_check.py

```diff
@@ -0,0 +1,179 @@
+
+
+regex = re.compile(r'((?P<hours>\d+?)hr)?((?P<minutes>\d+?)m)?((?P<seconds>\d+?)s)?')
+
+SLACheckBlueprint = Blueprint(
+    "sla_check", __name__,
+    template_folder='templates',
+    static_folder='static',
+    static_url_path='/static/sla_check')
+
+
+# Convenience method for creating the Flask response
+def get_base_response(status="OK", call_time=datetime.now(), include_arguments=True):
+  base_response = {"status": status, "call_time": call_time}
+  if include_arguments:
+    base_response["arguments"] = request.args
+  return base_response
+
+# Convenience method for finishing the Flask response
+def get_final_response(base_response, output=None, airflow_cmd=None):
+  final_response = base_response
+  final_response["response_time"] = datetime.now()
+  if output:
+    final_response["output"] = output
+  if airflow_cmd:
+    final_response["airflow_cmd"] = airflow_cmd
+  return jsonify(final_response)
+
+# Attempts to load a module from a Python file
+def load_dag_module(module_file):
+  module_folder = os.path.dirname(module_file)
+  module_file = os.path.basename(module_file)
+  module_name = os.path.splitext(module_file)[0]
+
+  fp, pathname, description = imp.find_module(module_name, [module_folder])
+
+  try:
+    return imp.load_module(module_name, fp, pathname, description)
+  except ImportError:
+    logging.error("Failed to load module %s", fp)
+  finally:
+    # Since we may exit via an exception, close fp explicitly.
+    if fp:
+      fp.close()
+
+# Loads all of the existing dags in Airflow and compiles a list
+# of all the dag names and sla values for DAG's that have a
+# DAG_SLA defined. DAG modules that cannot be loaded will be skipped.
+def find_slas():
+  dagbag = DagBag()
+  parent_dags = filter(lambda d: d.parent_dag is None, dagbag.dags.values())
+
+  output = []
+  for dag in parent_dags:
+    dag_module = load_dag_module(dag.full_filepath)
+
+    if dag_module and hasattr(dag_module, 'DAG_SLA'):
+      sla = dag_module.DAG_SLA
+      name = dag_module.DAG_NAME
+
+      output.append((name,sla))
+  return output
+
+# Computes the datetime today that yesterday's DAG execution is due to
+# be completed by.
+def calculate_sla_due(sla_str):
+  sla_delta = parse_timedelta(sla_str)
+
+  return datetime.now().replace(hour=0, minute=0, second=0, microsecond=0) + sla_delta
+
+# Takes in a dag_id and an SLA value and computes if the DAG is due now
+# and if it is due now, checks the latest dag_run and reports out of compliance
+# if it does not yet exist for yesterday or if it is still running.
+def check_compliance(dag_name, sla):
+  sla_due = calculate_sla_due(sla)
+
+  if datetime.now() < sla_due:
+    return True
+
+  yesterday = date.today() - timedelta(days=1)
+  dt = yesterday.strftime("%Y-%m-%d")
+
+  last_run = settings.Session.query(DagRun).filter(
+    DagRun.dag_id == dag_name,
+    DagRun.execution_date == dt
+  ).first()
+
+  return last_run and last_run.state != State.RUNNING
```

Comment:
what if the state is FAILED or QUEUED and we're over time.  does that mean we are meeting SLA? (it doesn't)